File: DynamicProxyConfiguration/testload/computeResults.py
#!/usr/bin/python -O
# -*- coding:utf8 -*-

# This script SHOULD NOT be called directly. It will be called by the makePlot.sh script.
# The input JSON files are the output of vegeta.

import os
import sys
import glob
import json

def usage():
    print >> sys.stderr, 'Usage: {0} <json files>'.format(sys.argv[0])

def parseResultsFilename(filename):
    fSplit = os.path.basename(filename).split('_')
    return {
        'xp_type': fSplit[0],
        'front': fSplit[1],
        'web_protocol': fSplit[2],
        'rate': fSplit[3],
        'duration': fSplit[4],
        'run': fSplit[5],
    }

if __name__ == '__main__':
    if len(sys.argv) < 2: 
        usage()
        sys.exit(-1)
    sys.argv.pop(0)

    latencies = {}

    for f in sys.argv:
        with open(f) as fp:
            results = json.load(fp)
            if results['errors'] is None:
                results['errors'] = []
        
        fSplit = os.path.basename(f).split('_')
        if fSplit[3] == '1500':
            continue
        hashKey = '_'.join(fSplit[0:-1])
        if hashKey not in latencies:
            latencies[hashKey] = {}
        latencies[hashKey]['mean'] = latencies[hashKey].get('mean', 0) + results['latencies']['mean']
        latencies[hashKey]['99th'] = latencies[hashKey].get('99th', 0) + results['latencies']['99th']
        latencies[hashKey]['nbErrors'] = 0
        for statusCode in results['status_codes']:
                if statusCode != '200':
                    latencies[hashKey]['nbErrors'] += results['status_codes'][statusCode]

        latencies[hashKey]['nbRun'] = latencies[hashKey].get('nbRun', 0) + 1

    # Purge old results files
    for name in glob.glob('results/*.dat'):
        os.remove(name)

    # Print the results and save them in a file for further plot
    for xp in sorted(latencies.keys()):
        print xp
        latencies[xp]['mean'] = latencies[xp]['mean'] / latencies[xp]['nbRun']
        latencies[xp]['99th'] = latencies[xp]['99th'] / latencies[xp]['nbRun']
        latencies[xp]['nbErrors'] = latencies[xp]['nbErrors'] / latencies[xp]['nbRun']
        # Results are in nanoseconds
        print 'mean: {0} ms\n99th: {1} ms\n# errors: {2}\n'.format(latencies[xp]['mean']*10e-7,
                latencies[xp]['99th']*10e-7, latencies[xp]['nbErrors'])
        xpSplit = xp.split('_')
        dataFilename = './results/' + '_'.join(xpSplit[0:3]) + '_' + xpSplit[-1] + ".dat"
        with open(dataFilename, 'a') as dataFile:
            dataFile.write("{}\t{}\t{}\n".format(xpSplit[3], latencies[xp]['mean']*10e-7,
                latencies[xp]['99th']*10e-7))

File: ContainersScheduling/compute_results.rb
#!/usr/bin/env ruby
#
# This script SHOULD NOT be called directly. It will be called by the makePlot.sh script.

require 'json'
require 'fileutils'

if ARGV.count.zero?
  STDERR.puts "Usage: #{$PROGRAM_NAME} <json files>"
  exit
end

results_dir = 'results'
scheduler_hosts = %w[ovh-sbg-hos03.staging.scalingo.com ovh-sbg-hos04.staging.scalingo.com]
latencies = {}

ARGV.each do |json_file|
  file = File.read json_file
  results = JSON.parse file

  fname_split = File.basename(json_file).split '_'
  app_name = fname_split[-1].split('.')[0]
  next if app_name == 'biniou'
  # hash_key is something like "scheduling_memory_60_run1_biniou-01"
  hash_key = "#{fname_split[0..-2].join('_')}_#{app_name}"
  latencies[hash_key] = Hash.new(0) if !latencies.key? hash_key

  latencies[hash_key]['appName'] = app_name
  latencies[hash_key]['mean'] = latencies[hash_key]['mean'] + results['latencies']['mean']
  latencies[hash_key]['99th'] = latencies[hash_key]['99th'] + results['latencies']['99th']
  latencies[hash_key]['nbErrors'] = 0
  results['status_codes'].reject do |status_code|
    status_code == '200'
  end.each_value do |amount|
    latencies[hash_key]['nbErrors'] += amount
  end
  latencies[hash_key]['nbRun'] += 1

  # Filter the scheduler log file starting from the latest "strategy=..." to the end of the file.
  scheduler_hosts.each do |host|
    host_id = host.split('.')[0].split('-')[-1] # e.g. hos01
    logfile_base = json_file.split('_')[0..-2].join('_')
    log_file = "#{logfile_base}_#{host_id}_app-scheduler.log"
    filtered_log_file = "#{logfile_base}_#{host_id}_app-scheduler_filtered.log"

    next if File.exist? filtered_log_file

    if !File.exist? log_file
      STDERR.puts "Missing log file (#{File.basename log_file}). Skip to the next experiment."
      next
    end

    rx = Regexp.new('.*(time=.* strategy=.*)', Regexp::MULTILINE)
    File.open(filtered_log_file, 'w') do |filtered_log|
      filtered_log.write(rx.match(File.read(log_file))[1])
    end

    # Remove some useles error message to make it easier to read by a human eye
    filtered_lines = File.new(filtered_log_file)
                         .readlines
                         .grep_v(/error monitoring the resources/)
                         .grep_v(/TOPIC_NOT_FOUND/)
    File.open(filtered_log_file, 'w') do |filtered_log|
      filtered_lines.each do |line|
        filtered_log.write(line)
      end
    end
  end

  # Test if the XP is valid
  nb_filtered_log_files = `ls -l #{results_dir}/#{fname_split[0..-2].join('_')}_*_app-scheduler_filtered.log | wc -l`.strip.to_i
  if nb_filtered_log_files != 2
    STDERR.puts "Lack a filtered log file (#{fname_split[0..-2].join('_')})"
    next
  end
  nb_scheduled_container = `cat #{results_dir}/#{fname_split[0..-2].join('_')}_*_app-scheduler_filtered.log | grep "_event=start-container container=biniou/" | wc -l`.strip.to_i
  if nb_scheduled_container != 4
    STDERR.puts "#{results_dir}/#{fname_split[0..-2].join('_')}"
    STDERR.puts "Wrong number of scheduled containers (#{nb_scheduled_container} != 4)"
    next
  end
end

# Remove any existing dat file
latencies.each do |xp, _|
  data_filename = xp.split('_')[0..-2].join('_')
  data_file = "#{results_dir}/#{data_filename}.dat"
  FileUtils.remove_file data_file if File.exist? data_file
end

# Print the results and save them in a file for further plot
latencies.each do |xp, results|
  # puts "XP: #{xp}"

  # Results are in nanoseconds, convert to miliseconds
  results['mean'] = results['mean'] / results['nbRun'] * 10e-7
  results['99th'] = results['99th'] / results['nbRun'] * 10e-7
  results['nbErrors'] = results['nbErrors'] / results['nbRun']
  # puts "mean: #{results['mean']} ms\n99th: #{results['99th']} ms\n# errors: #{results['nbErrors']}"

  data_filename = xp.split('_')[0..-2].join('_')
  data_file = "#{results_dir}/#{data_filename}.dat"

  File.open(data_file, 'a') do |f|
    f.write(results.each_value.to_a.join("\t") + "\n")
  end
end
